src/expr/impl/src/scalar/date_bin.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chrono::DateTime;
+use risingwave_common::types::{Interval, Timestamp, Timestamptz};
+use risingwave_expr::{ExprError, Result, function};
+
+#[function("date_bin(interval, timestamp, timestamp) -> timestamp")]
+pub fn date_bin_ts(stride: Interval, source: Timestamp, origin: Timestamp) -> Result<Timestamp> {
+    let source_us = source.0.and_utc().timestamp_micros(); // source to microseconds
+    let origin_us = origin.0.and_utc().timestamp_micros(); // origin to microseconds
+
+    let binned_source_us = date_bin_inner(stride, source_us, origin_us)?;
+    Ok(Timestamp(
+        DateTime::from_timestamp_micros(binned_source_us)
+            .unwrap()
+            .naive_utc(),
+    ))
+}
+
+#[function("date_bin(interval, timestamptz, timestamptz) -> timestamptz")]
+pub fn date_bin_tstz(
+    stride: Interval,
+    source: Timestamptz,
+    origin: Timestamptz,
+) -> Result<Timestamptz> {
+    let source_us = source.timestamp_micros(); // source to microseconds
+    let origin_us = origin.timestamp_micros(); // origin to microseconds
+
+    let binned_source_us = date_bin_inner(stride, source_us, origin_us)?;
+    Ok(Timestamptz::from_micros(binned_source_us))
+}
+
+#[function("date_bin(interval, timestamp, timestamptz) -> timestamptz")]
+pub fn date_bin_ts_tstz(
+    stride: Interval,
+    source: Timestamp,
+    origin: Timestamptz,
+) -> Result<Timestamptz> {
+    let source_us = source.0.and_utc().timestamp_micros(); // source to microseconds
+    let origin_us = origin.timestamp_micros(); // origin to microseconds
```

Comment:
Mixed usage of timestamp and timestamptz is wrong. PostgreSQL only defines 2 overloads: `select * from pg_proc where proname = 'date_bin';`

When the input is mixed, PostgreSQL (and RisingWave) tries to add implicit cast to find a best match (#3393). In this case, the timestamp argument will be casted into timestamptz **in the session zone** (NOT `0.and_utc()`).

The fix is also simple: just remove the implementation of these 2 overloads.

Reply:
Updated. Thanks!